File: clients/php/src/Watermark.php
<?php

declare(strict_types=1);

namespace Sharptown\Client;

/**
 * An image overlay composited onto the result. Build it from a URL the server fetches
 * ({@link url()}), or from local bytes/a file uploaded with the request ({@link bytes()},
 * {@link file()}), then chain the placement and appearance methods.
 *
 * @example
 * $st->transform($file)
 *     ->composite(Watermark::url('https://cdn.example.com/logo.png')->resize(120)->opacity(0.6))
 *     ->convert('webp');
 */
final class Watermark implements CompositeMark
{
    /** @var array<string, mixed> */
    private array $spec = ['type' => 'image'];
    private ?string $upload = null;

    private function __construct()
    {
    }

    /** An image watermark fetched from $url by the server. */
    public static function url(string $url): self
    {
        $mark = new self();
        $mark->spec['url'] = $url;
        return $mark;
    }

    /** An image watermark uploaded from raw bytes. */
    public static function bytes(string $data): self
    {
        $mark = new self();
        $mark->upload = $data;
        return $mark;
    }

    /** An image watermark uploaded from a local file path. */
    public static function file(string $path): self
    {
        $data = @file_get_contents($path);
        if ($data === false) {
            throw new SharptownError(sprintf('Watermark::file: cannot read %s', $path));
        }
        return self::bytes($data);
    }

    /** Fits the overlay inside $width×$height. Either dimension is optional. */
    public function resize(?int $width = null, ?int $height = null): self
    {
        if ($width !== null) {
            $this->spec['width'] = Operations::toPositiveInt($width, 'watermark width');
        }
        if ($height !== null) {
            $this->spec['height'] = Operations::toPositiveInt($height, 'watermark height');
        }
        return $this;
    }

    /** Sets the overlay width only. */
    public function width(int $value): self
    {
        $this->spec['width'] = Operations::toPositiveInt($value, 'watermark width');
        return $this;
    }

    /** Sets the overlay height only. */
    public function height(int $value): self
    {
        $this->spec['height'] = Operations::toPositiveInt($value, 'watermark height');
        return $this;
    }

    /** Rotates the overlay by degrees. */
    public function rotate(int $degrees): self
    {
        $this->spec['rotate'] = $degrees;
        return $this;
    }

    /** Sets the overlay opacity (0–1). */
    public function opacity(int|float $value): self
    {
        $this->spec['opacity'] = Operations::toRange($value, 'watermark opacity', 0, 1);
        return $this;
    }

    /** Sets the placement gravity (default southeast). */
    public function gravity(string $value): self
    {
        $this->spec['gravity'] = $value;
        return $this;
    }

    /** Places the overlay at ($x, $y) from the top-left instead of a gravity. */
    public function offset(int $x, int $y): self
    {
        $this->spec['x'] = $x;
        $this->spec['y'] = $y;
        return $this;
    }

    /** Repeats the overlay across the whole image. */
    public function tile(bool $enabled = true): self
    {
        $this->spec['tile'] = $enabled;
        return $this;
    }

    /** Sets the Sharp blend mode (default over). */
    public function blend(string $mode): self
    {
        $this->spec['blend'] = $mode;
        return $this;
    }

    public function resolve(): array
    {
        return ['spec' => $this->spec, 'bytes' => $this->upload];
    }
}

File: clients/go/watermark.go
package sharptown

// CompositeMark is an overlay passed to (*Transform).Composite — either a Watermark (image)
// or a Textmark (text). resolve returns its wire spec plus optional bytes to upload.
type CompositeMark interface {
	resolve() (map[string]any, []byte)
}

// Watermark is an image overlay composited onto the result. Create it with NewWatermark (a
// URL the server fetches) or NewWatermarkBytes (uploaded with the request), then chain the
// placement and appearance methods.
type Watermark struct {
	spec  map[string]any
	bytes []byte
}

// NewWatermark builds an image watermark fetched from url by the server.
func NewWatermark(url string) *Watermark {
	return &Watermark{spec: map[string]any{"type": "image", "url": url}}
}

// NewWatermarkBytes builds an image watermark uploaded alongside the request.
func NewWatermarkBytes(data []byte) *Watermark {
	return &Watermark{spec: map[string]any{"type": "image"}, bytes: data}
}

// Resize fits the overlay inside width×height. A non-positive dimension is left unset.
func (w *Watermark) Resize(width, height int) *Watermark {
	if width > 0 {
		w.spec["width"] = width
	}
	if height > 0 {
		w.spec["height"] = height
	}
	return w
}

// Width sets the overlay width only.
func (w *Watermark) Width(value int) *Watermark { w.spec["width"] = value; return w }

// Height sets the overlay height only.
func (w *Watermark) Height(value int) *Watermark { w.spec["height"] = value; return w }

// Rotate rotates the overlay by degrees.
func (w *Watermark) Rotate(degrees int) *Watermark { w.spec["rotate"] = degrees; return w }

// Opacity sets the overlay opacity (0–1).
func (w *Watermark) Opacity(value float64) *Watermark { w.spec["opacity"] = value; return w }

// Gravity sets the placement gravity (default southeast).
func (w *Watermark) Gravity(value string) *Watermark { w.spec["gravity"] = value; return w }

// Offset places the overlay at (x, y) from the top-left instead of a gravity.
func (w *Watermark) Offset(x, y int) *Watermark { w.spec["x"] = x; w.spec["y"] = y; return w }

// Tile repeats the overlay across the whole image.
func (w *Watermark) Tile() *Watermark { w.spec["tile"] = true; return w }

// Blend sets the Sharp blend mode (default over).
func (w *Watermark) Blend(mode string) *Watermark { w.spec["blend"] = mode; return w }

func (w *Watermark) resolve() (map[string]any, []byte) { return w.spec, w.bytes }

// Textmark is a text overlay composited onto the result, rendered server-side.
type Textmark struct {
	spec map[string]any
}

// NewTextmark builds a text watermark.
func NewTextmark(text string) *Textmark {
	return &Textmark{spec: map[string]any{"type": "text", "text": text}}
}

// Size sets the font size in pixels.
func (m *Textmark) Size(value int) *Textmark { m.spec["size"] = value; return m }

// Color sets the text colour (any CSS colour).
func (m *Textmark) Color(value string) *Textmark { m.spec["color"] = value; return m }

// Font sets the font family.
func (m *Textmark) Font(value string) *Textmark { m.spec["font"] = value; return m }

// Weight sets the font weight (e.g. "bold").
func (m *Textmark) Weight(value string) *Textmark { m.spec["weight"] = value; return m }

// Background paints a colour behind the text tile.
func (m *Textmark) Background(value string) *Textmark { m.spec["background"] = value; return m }

// Rotate rotates the text by degrees.
func (m *Textmark) Rotate(degrees int) *Textmark { m.spec["rotate"] = degrees; return m }

// Opacity sets the text opacity (0–1).
func (m *Textmark) Opacity(value float64) *Textmark { m.spec["opacity"] = value; return m }

// Gravity sets the placement gravity.
func (m *Textmark) Gravity(value string) *Textmark { m.spec["gravity"] = value; return m }

// Offset places the text at (x, y) from the top-left.
func (m *Textmark) Offset(x, y int) *Textmark { m.spec["x"] = x; m.spec["y"] = y; return m }

// Tile repeats the text across the whole image.
func (m *Textmark) Tile() *Textmark { m.spec["tile"] = true; return m }

func (m *Textmark) resolve() (map[string]any, []byte) { return m.spec, nil }
